core/activity/src/requestor/control.rs

```diff
@@ -15,51 +19,99 @@ use crate::common::{
     QueryTimeout, QueryTimeoutCommandIndex,
 };
 use crate::dao::ActivityDao;
-use crate::error::Error;
+use crate::{error::Error, Result};
 
 pub fn extend_web_scope(scope: actix_web::Scope) -> actix_web::Scope {
     scope
         .service(create_activity)
         .service(destroy_activity)
         .service(exec)
         .service(get_batch_results)
+        .service(encrypted)
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum CreateActivityJson {
+    Agreement(String),
+    Request(CreateActivityRequest),
+}
+
+impl CreateActivityJson {
+    fn agreement_id(&self) -> &str {
+        match self {
+            Self::Agreement(agreement_id) => agreement_id.as_ref(),
+            Self::Request(request) => request.agreement_id.as_ref(),
+        }
+    }
+
+    fn pub_key(&self) -> Result<Option<Vec<u8>>> {
+        match self {
+            Self::Request(CreateActivityRequest {
+                requestor_pub_key: Some(pub_key),
+                ..
+            }) => {
+                let bytes = hex::decode(pub_key)
+                    .map_err(|e| Error::BadRequest(format!("Invalid requestor pub key: {}", e)))?;
+
+                Ok(Some(bytes))
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn to_response(&self, result: CreateActivityResult) -> serde_json::Value {
+        match self {
+            Self::Agreement(_) => serde_json::json!(result.activity_id),
+            Self::Request(_) => serde_json::json!(result),
+        }
+    }
 }
 
 /// Creates new Activity based on given Agreement.
 #[actix_web::post("/activity")]
 async fn create_activity(
     db: web::Data<DbExecutor>,
     query: web::Query<QueryTimeout>,
-    body: web::Json<String>,
+    body: web::Json<CreateActivityJson>,
     id: Identity,
 ) -> impl Responder {
-    let agreement_id = body.into_inner();
-    authorize_agreement_initiator(id.identity, &agreement_id).await?;
+    let agreement_id = body.agreement_id();
+    authorize_agreement_initiator(id.identity, agreement_id).await?;
 
     let agreement = get_agreement(&agreement_id).await?;
-    log::trace!("agreement: {:#?}", agreement);
+    log::info!("agreement: {:#?}", agreement);
```

Comment:
debug!

Reply:
I would even leave it at trace level. I don't think we should log whole structures on debug level, because it prevents us from using this level..